horovod/common/ops/gpu_operations.cc

```diff
@@ -213,6 +216,68 @@ void GPUAllreduce::MemcpyInFusionBuffer(
 }
 #endif
 
+#if HAVE_ROCM
```

Comment:
In this file there are now four member functions `MemcpyInFusionBuffer`, `ScaleMemcpyInFusionBuffer`, `MemcpyOutFusionBuffer`, and `ScaleMemcpyOutFusionBuffer` that are almost identical for `HAVE_CUDA` and `HAVE_ROCM`.

For each of these, I think it would be better to have more local `#if` guards and avoid a lot of repetition. So for instance there could be just one definition of `GPUAllreduce::MemcpyInFusionBuffer` and inside we could have
```
#if HAVE_CUDA
        BatchedD2DMemcpyCudaImpl(
            d2d_params, count,
            gpu_context_->streams[global_state_->current_nccl_stream]
                                 [first_entry.device]);
#elif HAVE_ROCM
        BatchedD2DMemcpyROCmImpl(
            d2d_params, count,
            gpu_context_->streams[global_state_->current_nccl_stream]
                                 [first_entry.device]);
#endif
```
The overall function would then be guarded by
```
#if HAVE_CUDA || HAVE_ROCM
void GPUAllreduce::MemcpyInFusionBuffer(
    const std::vector<TensorTableEntry>& entries, const void*& fused_input_data,
    void*& buffer_data, size_t& buffer_len) {
// ...
#endif
```

I think we should go for this very simple unification already in this PR.

Reply:
This is taken care of in the latest commit.